File: mobile/ios/IwanyuNative/Sources/SupabaseAPI.swift
import Foundation

enum APIError: Error {
    case missingConfig
    case invalidResponse
    case status(Int, String)
}

struct AuthSession: Codable {
    let access_token: String
    let refresh_token: String?
    let user: AuthUser
}

struct AuthUser: Codable {
    let id: String
    let email: String?
}

struct FunctionResult: Codable {
    let success: Bool?
    let orderId: String?
    let total: Int?
    let message: String?
}

final class SupabaseAPI {
    private let baseURL: URL
    private let anonKey: String

    init() {
        guard
            let urlString = Bundle.main.object(forInfoDictionaryKey: "SUPABASE_URL") as? String,
            let url = URL(string: urlString),
            let key = Bundle.main.object(forInfoDictionaryKey: "SUPABASE_ANON_KEY") as? String,
            !key.isEmpty
        else {
            fatalError("Missing SUPABASE_URL / SUPABASE_ANON_KEY in Info.plist")
        }
        self.baseURL = url
        self.anonKey = key
    }

    // MARK: - Auth (same Supabase backend)

    func signIn(email: String, password: String) async throws -> AuthSession {
        let payload: [String: Any] = ["email": email, "password": password]
        let data = try await postRaw(path: "/auth/v1/token?grant_type=password", payload: payload, bearer: anonKey)
        return try JSONDecoder().decode(AuthSession.self, from: data)
    }

    func signUp(email: String, password: String) async throws -> AuthSession {
        let payload: [String: Any] = ["email": email, "password": password]
        let data = try await postRaw(path: "/auth/v1/signup", payload: payload, bearer: anonKey)
        return try JSONDecoder().decode(AuthSession.self, from: data)
    }

    // MARK: - Read APIs

    func fetchProducts(limit: Int) async throws -> [Product] {
        let path = "/rest/v1/products?select=id,vendor_id,title,description,category,price_rwf,image_url,in_stock,free_shipping,rating,review_count,discount_percentage&deleted_at=is.null&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: anonKey)
    }

    func fetchVendors(limit: Int) async throws -> [Vendor] {
        let path = "/rest/v1/vendors?select=id,name,location,verified,status,owner_user_id&deleted_at=is.null&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: anonKey)
    }

    func fetchProfileRole(userId: String, token: String) async throws -> String {
        let path = "/rest/v1/profiles?select=role&id=eq.\(userId)&limit=1"
        let rows: [ProfileRoleRow] = try await get(path: path, token: token)
        return rows.first?.role ?? "buyer"
    }

    func fetchVendorsForOwner(userId: String, token: String) async throws -> [Vendor] {
        let path = "/rest/v1/vendors?select=id,name,location,verified,status,owner_user_id&owner_user_id=eq.\(userId)&deleted_at=is.null&limit=50"
        return try await get(path: path, token: token)
    }

    func fetchSellerOrderItems(vendorIds: [String], token: String, limit: Int = 100) async throws -> [SellerOrderItem] {
        guard !vendorIds.isEmpty else { return [] }
        let inList = vendorIds.joined(separator: ",")
        let path = "/rest/v1/order_items?select=order_id,product_id,vendor_id,title,quantity,status&vendor_id=in.(\(inList))&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: token)
    }

    func fetchSellerPayouts(vendorIds: [String], token: String, limit: Int = 100) async throws -> [VendorPayout] {
        guard !vendorIds.isEmpty else { return [] }
        let inList = vendorIds.joined(separator: ",")
        let path = "/rest/v1/vendor_payouts?select=id,vendor_id,order_id,amount_rwf,status&vendor_id=in.(\(inList))&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: token)
    }

    func fetchAdminVendorApplications(token: String, limit: Int = 100) async throws -> [VendorApplication] {
        let path = "/rest/v1/vendor_applications?select=id,store_name,status,owner_user_id,created_at&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: token)
    }

    func fetchAdminDiscountCodes(token: String, limit: Int = 100) async throws -> [DiscountCode] {
        let path = "/rest/v1/discount_codes?select=id,code,discount_type,active,redeemed_count&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: token)
    }

    func fetchOrders(userId: String, token: String, limit: Int = 50) async throws -> [Order] {
        let path = "/rest/v1/orders?select=id,buyer_user_id,buyer_email,shipping_address,status,total_rwf,created_at&buyer_user_id=eq.\(userId)&order=created_at.desc&limit=\(limit)"
        return try await get(path: path, token: token)
    }

    func fetchWishlist(userId: String, token: String, limit: Int = 100) async throws -> [[String: String]] {
        let path = "/rest/v1/wishlist_items?select=product_id,user_id&user_id=eq.\(userId)&limit=\(limit)"
        return try await get(path: path, token: token)
    }

    // MARK: - Edge functions (same website APIs)

    func createOrder(accessToken: String, payload: [String: Any]) async throws -> FunctionResult {
        let data = try await postRaw(path: "/functions/v1/create-order", payload: payload, bearer: accessToken)
        return try JSONDecoder().decode(FunctionResult.self, from: data)
    }

    func verifyFlutterwavePayment(accessToken: String, orderId: String, transactionId: String) async throws -> FunctionResult {
        let payload: [String: Any] = ["orderId": orderId, "transactionId": transactionId]
        let data = try await postRaw(path: "/functions/v1/flutterwave-verify", payload: payload, bearer: accessToken)
        return try JSONDecoder().decode(FunctionResult.self, from: data)
    }

    func initFlutterwavePayment(accessToken: String, payload: [String: Any]) async throws -> [String: String] {
        let data = try await postRaw(path: "/functions/v1/flutterwave-init", payload: payload, bearer: accessToken)
        return (try JSONSerialization.jsonObject(with: data) as? [String: String]) ?? [:]
    }

    // MARK: - HTTP helpers

    private func get<T: Decodable>(path: String, token: String) async throws -> T {
        guard let url = URL(string: path, relativeTo: baseURL) else { throw APIError.invalidResponse }
        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(anonKey, forHTTPHeaderField: "apikey")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError.status(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func postRaw(path: String, payload: [String: Any], bearer: String) async throws -> Data {
        guard let url = URL(string: path, relativeTo: baseURL) else { throw APIError.invalidResponse }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(anonKey, forHTTPHeaderField: "apikey")
        req.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization")
        req.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError.status(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
        return data
    }
}

File: mobile/android/app/src/main/java/com/iwanyu/mobile/SupabaseApi.kt
package com.iwanyu.mobile

import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.OkHttpClient
import okhttp3.Request

@Serializable
data class Product(
    val id: String,
    val vendor_id: String,
    val title: String,
    val description: String? = null,
    val category: String? = null,
    val price_rwf: Int,
    val image_url: String? = null,
    val in_stock: Boolean? = null,
    val free_shipping: Boolean? = null,
    val rating: Double? = null,
    val review_count: Int? = null,
    val discount_percentage: Int? = null,
)

@Serializable
data class AuthUser(
    val id: String,
    val email: String? = null,
)

@Serializable
data class AuthSession(
    val access_token: String,
    val refresh_token: String? = null,
    val user: AuthUser,
)

@Serializable
data class FunctionResult(
    val success: Boolean? = null,
    val orderId: String? = null,
    val total: Int? = null,
    val message: String? = null,
)

object SupabaseApi {
    private val client = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    private val baseUrl = BuildConfig.SUPABASE_URL.trimEnd('/')
    private val anonKey = BuildConfig.SUPABASE_ANON_KEY

    fun fetchProducts(limit: Int): List<Product> {
        require(baseUrl.isNotBlank()) { "SUPABASE_URL is missing" }
        require(anonKey.isNotBlank()) { "SUPABASE_ANON_KEY is missing" }

        val url = "$baseUrl/rest/v1/products?select=id,vendor_id,title,description,category,price_rwf,image_url,in_stock,free_shipping,rating,review_count,discount_percentage&deleted_at=is.null&order=created_at.desc&limit=$limit"

        val req = Request.Builder()
            .url(url)
            .addHeader("apikey", anonKey)
            .addHeader("Authorization", "Bearer $anonKey")
            .addHeader("Content-Type", "application/json")
            .build()

        client.newCall(req).execute().use { res ->
            if (!res.isSuccessful) error("HTTP ${res.code}: ${res.body?.string()}")
            val body = res.body?.string().orEmpty()
            return json.decodeFromString(ListSerializer(Product.serializer()), body)
        }
    }

    fun fetchVendors(limit: Int = 100): List<Map<String, String>> {
        val url = "$baseUrl/rest/v1/vendors?select=id,name,location,verified,status&deleted_at=is.null&order=created_at.desc&limit=$limit"
        val req = baseGet(url)
        client.newCall(req).execute().use { res ->
            if (!res.isSuccessful) error("HTTP ${res.code}: ${res.body?.string()}")
            val body = res.body?.string().orEmpty()
            return json.decodeFromString(
                ListSerializer(MapSerializer(String.serializer(), String.serializer())),
                body
            )
        }
    }

    fun fetchOrders(userId: String, limit: Int = 50): List<Map<String, String>> {
        val url = "$baseUrl/rest/v1/orders?select=id,buyer_user_id,buyer_email,shipping_address,status,total_rwf,created_at&buyer_user_id=eq.$userId&order=created_at.desc&limit=$limit"
        val req = baseGet(url)
        client.newCall(req).execute().use { res ->
            if (!res.isSuccessful) error("HTTP ${res.code}: ${res.body?.string()}")
            val body = res.body?.string().orEmpty()
            return json.decodeFromString(
                ListSerializer(MapSerializer(String.serializer(), String.serializer())),
                body
            )
        }
    }

    fun fetchWishlist(userId: String, limit: Int = 100): List<Map<String, String>> {
        val url = "$baseUrl/rest/v1/wishlist_items?select=product_id,user_id&user_id=eq.$userId&limit=$limit"
        val req = baseGet(url)
        client.newCall(req).execute().use { res ->
            if (!res.isSuccessful) error("HTTP ${res.code}: ${res.body?.string()}")
            val body = res.body?.string().orEmpty()
            return json.decodeFromString(
                ListSerializer(MapSerializer(String.serializer(), String.serializer())),
                body
            )
        }
    }

    fun createOrder(accessToken: String, jsonPayload: String): String {
        return invokeFunction("create-order", accessToken, jsonPayload)
    }

    fun createOrderResult(accessToken: String, jsonPayload: String): FunctionResult {
        val raw = createOrder(accessToken, jsonPayload)
        return json.decodeFromString(FunctionResult.serializer(), raw)
    }

    fun verifyFlutterwavePayment(accessToken: String, orderId: String, transactionId: String): String {
        val payload = "{\"orderId\":\"$orderId\",\"transactionId\":\"$transactionId\"}"
        return invokeFunction("flutterwave-verify", accessToken, payload)
    }

    fun signIn(email: String, password: String): AuthSession {
        val payload = "{\"email\":\"$email\",\"password\":\"$password\"}"
        val url = "$baseUrl/auth/v1/token?grant_type=password"
        val req = Request.Builder()
            .url(url)
            .addHeader("apikey", anonKey)
            .addHeader("Authorization", "Bearer $anonKey")
            .addHeader("Content-Type", "application/json")
            .post(payload.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(req).execute().use { res ->
            val body = res.body?.string().orEmpty()
            if (!res.isSuccessful) error("HTTP ${res.code}: $body")
            return json.decodeFromString(AuthSession.serializer(), body)
        }
    }

    fun signUp(email: String, password: String): AuthSession {
        val payload = "{\"email\":\"$email\",\"password\":\"$password\"}"
        val url = "$baseUrl/auth/v1/signup"
        val req = Request.Builder()
            .url(url)
            .addHeader("apikey", anonKey)
            .addHeader("Authorization", "Bearer $anonKey")
            .addHeader("Content-Type", "application/json")
            .post(payload.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(req).execute().use { res ->
            val body = res.body?.string().orEmpty()
            if (!res.isSuccessful) error("HTTP ${res.code}: $body")
            return json.decodeFromString(AuthSession.serializer(), body)
        }
    }

    private fun invokeFunction(slug: String, accessToken: String, jsonPayload: String): String {
        val url = "$baseUrl/functions/v1/$slug"
        val req = Request.Builder()
            .url(url)
            .addHeader("apikey", anonKey)
            .addHeader("Authorization", "Bearer $accessToken")
            .addHeader("Content-Type", "application/json")
            .post(jsonPayload.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(req).execute().use { res ->
            val body = res.body?.string().orEmpty()
            if (!res.isSuccessful) error("HTTP ${res.code}: $body")
            return body
        }
    }

    private fun baseGet(url: String): Request {
        return Request.Builder()
            .url(url)
            .addHeader("apikey", anonKey)
            .addHeader("Authorization", "Bearer $anonKey")
            .addHeader("Content-Type", "application/json")
            .build()
    }
}
